File: csharp/rocketmq-client-csharp/SimpleSubscriptionSettings.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Google.Protobuf.WellKnownTypes;
using NLog;
using Proto = Apache.Rocketmq.V2;

namespace Org.Apache.Rocketmq
{
    public class SimpleSubscriptionSettings : Settings
    {
        private static readonly Logger Logger = MqLogManager.Instance.GetCurrentClassLogger();

        private readonly Resource _group;
        private readonly TimeSpan _longPollingTimeout;
        private readonly ConcurrentDictionary<string /* topic */, FilterExpression> _subscriptionExpressions;

        public SimpleSubscriptionSettings(string clientId, Endpoints endpoints, string consumerGroup,
            TimeSpan requestTimeout, TimeSpan longPollingTimeout,
            ConcurrentDictionary<string, FilterExpression> subscriptionExpressions) : base(
            clientId, ClientType.SimpleConsumer, endpoints, requestTimeout)
        {
            _group = new Resource(consumerGroup);
            _longPollingTimeout = longPollingTimeout;
            _subscriptionExpressions = subscriptionExpressions;
        }

        public override void Sync(Proto::Settings settings)
        {
            if (Proto.Settings.PubSubOneofCase.Subscription != settings.PubSubCase)
            {
                Logger.Error($"[Bug] Issued settings doesn't match with the client type, clientId={ClientId}, " +
                             $"pubSubCase={settings.PubSubCase}, clientType={ClientType}");
            }
        }

        public override Proto.Settings ToProtobuf()
        {
            var subscriptionEntries = new List<Proto.SubscriptionEntry>();
            foreach (var (key, value) in _subscriptionExpressions)
            {
                var topic = new Proto.Resource()
                {
                    Name = key,
                };
                var subscriptionEntry = new Proto.SubscriptionEntry();
                var filterExpression = new Proto.FilterExpression();
                switch (value.Type)
                {
                    case ExpressionType.Tag:
                        filterExpression.Type = Proto.FilterType.Tag;
                        break;
                    case ExpressionType.Sql92:
                        filterExpression.Type = Proto.FilterType.Sql;
                        break;
                    default:
                        Logger.Warn($"[Bug] Unrecognized filter type={value.Type} for simple consumer");
                        break;
                }

                filterExpression.Expression = value.Expression;
                subscriptionEntry.Topic = topic;
                subscriptionEntries.Add(subscriptionEntry);
            }

            var subscription = new Proto.Subscription
            {
                Group = _group.ToProtobuf(),
                Subscriptions = { subscriptionEntries },
                LongPollingTimeout = Duration.FromTimeSpan(_longPollingTimeout)
            };
            return new Proto.Settings
            {
                AccessPoint = Endpoints.ToProtobuf(),
                ClientType = ClientTypeHelper.ToProtobuf(ClientType),
                RequestTimeout = Duration.FromTimeSpan(RequestTimeout),
                Subscription = subscription,
                UserAgent = UserAgent.Instance.ToProtobuf()
            };
        }
    }
}

File: python/rocketmq/simple_subscription_settings.py

from typing import Dict
from rocketmq.filter_expression import ExpressionType
from rocketmq.log import logger

# Assuming these imports based on the provided context
from rocketmq.protocol.definition_pb2 import (
    Settings as ProtoSettings,
    SubscriptionEntry as ProtoSubscriptionEntry,
    Resource as ProtoResource,
    FilterExpression as ProtoFilterExpression,
    FilterType as ProtoFilterType,
    Subscription as ProtoSubscription
)
from google.protobuf.duration_pb2 import Duration

# Your provided Enums and classes
from .settings import ClientType, ClientTypeHelper, Settings


# Assuming a simple representation of FilterExpression for the purpose of this example
class FilterExpression:
    def __init__(self, type, expression):
        self.Type = type
        self.Expression = expression


class SimpleSubscriptionSettings(Settings):

    def __init__(self, clientId, endpoints, consumerGroup, requestTimeout, longPollingTimeout,
                 subscriptionExpressions: Dict[str, FilterExpression]):
        super().__init__(clientId, ClientType.SimpleConsumer, endpoints, None, requestTimeout)
        self._group = consumerGroup  # Simplified as string for now
        self._longPollingTimeout = longPollingTimeout
        self._subscriptionExpressions = subscriptionExpressions

    def Sync(self, settings: ProtoSettings):
        if not isinstance(settings, ProtoSettings):
            logger.error(f"[Bug] Issued settings doesn't match with the client type, clientId={self.ClientId}, clientType={self.ClientType}")

    def to_protobuf(self):
        subscriptionEntries = []

        for key, value in self._subscriptionExpressions.items():
            topic = ProtoResource()
            topic.name = key

            subscriptionEntry = ProtoSubscriptionEntry()
            filterExpression = ProtoFilterExpression()

            if value.type == ExpressionType.Tag:
                filterExpression.type = ProtoFilterType.TAG
            elif value.type == ExpressionType.Sql92:
                filterExpression.type = ProtoFilterType.SQL
            else:
                logger.warn(f"[Bug] Unrecognized filter type={value.Type} for simple consumer")

            filterExpression.expression = value.expression
            subscriptionEntry.topic.CopyFrom(topic)
            subscriptionEntries.append(subscriptionEntry)

        subscription = ProtoSubscription()
        group = ProtoResource()
        group.name = self._group
        subscription.group.CopyFrom(group)
        subscription.subscriptions.extend(subscriptionEntries)
        duration_longPollingTimeout = Duration(seconds=self._longPollingTimeout)
        subscription.long_polling_timeout.CopyFrom(duration_longPollingTimeout)

        settings = super().to_protobuf()
        settings.access_point.CopyFrom(self.Endpoints.to_protobuf())  # Assuming Endpoints has a to_protobuf method
        settings.client_type = ClientTypeHelper.to_protobuf(self.ClientType)

        settings.request_timeout.CopyFrom(Duration(seconds=int(self.RequestTimeout.total_seconds())))
        settings.subscription.CopyFrom(subscription)

        return settings
